discovery-provider/integration_tests/tasks/test_index_trending_underground_notification.py

```diff
@@ -0,0 +1,182 @@
+        )
+        trending_underground_tracks = helpers.query_result_to_list(
+            trending_underground_tracks
+        )
+
+        trending_underground_tracks_map = dict(
+            zip(
+                [track["track_id"] for track in trending_underground_tracks],
+                trending_underground_tracks,
+            )
+        )
+
+        trending_underground_tracks = [
+            trending_underground_tracks_map[track_id] for track_id in track_ids
+        ]
+
+        trending_underground_tracks = (trending_underground_tracks, track_ids)
+        set_json_cached_key(redis, trending_key, trending_underground_tracks)
+
+
+def test_index_trending_underground_notification(app, caplog):
+    """
+    Test that underground notifications are correctly calculated
+    based on current trending standings
+    """
+    caplog.set_level(logging.WARNING)
+
+    with app.app_context():
+        db = get_db()
+        # Add some users to the db so we have blocks
+        entities = {
+            "tracks": [
+                {
+                    "track_id": i,
+                    "owner_id": i % 2 + 1,
+                }
+                for i in range(100)
+            ],
+            "users": [{}] * 3,
+        }
+        populate_mock_db(db, entities)
+
+        # Test that if there are no prev notifications, all trending in top are generated
+        index_underground_trending_mock(db, list(range(1, 21)))
+        base_time_int = int(round(BASE_TIME.timestamp()))
+        index_trending_underground_notifications(db, base_time_int)
+
+        with db.scoped_session() as session:
+            notifications = (
+                session.query(Notification).order_by(asc(Notification.specifier)).all()
+            )
+            assert len(notifications) == 5
+
+            notification_0 = notifications[0]
+            assert notification_0.specifier == "1"
+            assert (
+                notification_0.group_id
+                == "trending_underground:time_range:week:genre:all:rank:1:track_id:1:timestamp:1672531200"
+            )
+            assert notification_0.type == "trending_underground"
+            assert notification_0.data == {
+                "rank": 1,
+                "genre": "all",
+                "track_id": 1,
+                "time_range": "week",
+            }
+            assert notification_0.user_ids == [2]
+
+            notification_1 = notifications[1]
+            assert notification_1.specifier == "2"
+            assert (
+                notification_1.group_id
+                == "trending_underground:time_range:week:genre:all:rank:2:track_id:2:timestamp:1672531200"
+            )
+            assert notification_1.type == "trending_underground"
+            assert notification_1.data == {
+                "rank": 2,
+                "genre": "all",
+                "track_id": 2,
+                "time_range": "week",
+            }
+            assert notification_1.user_ids == [1]
```

Comment:
nit / up to you - these assertions could be dry'd up with a for loop something like
`for idx, notification in enumerate(notifications):
 assert notification.type == "trending_underground"
            assert notification.data == {
                "rank": i+1,
                "genre": "all",
                "track_id": i+1,
                "time_range": "week",
            }
            assert notification_1.user_ids == [i % 2 + 1]
`

Reply:
totally agreed ^^ id like to do a pass on all the testing for indexing maybe once playlist indexing is done so i can see what parts are the same? im just a bit worried with my current python skills not to do too much this first go around. i absolutely want to make the testing suite solid though before considering this project done though.

Reply:
i am def going to add the enumeration part for these asserts though, great idea